File: php/socket_chat_server.php
<?php
error_reporting(E_ALL);
$port = 9050;
$sock = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
socket_set_option($sock, SOL_SOCKET, SO_REUSEADDR, 1);
socket_bind($sock, 0, $port);
socket_listen($sock);

$clients = array($sock);
$auth_clients = array();
while (true) {

    $read = $clients;
    if (socket_select($read, $write = NULL, $except = NULL, 10) < 1)
        continue;

    if (in_array($sock, $read)) {
        $new_sock = socket_accept($sock);
        $clients[] = $new_sock;
        socket_write($new_sock, "This simple chat server. Type 'iam:username' for authorize!\n");
        echo "There are " . (count($clients) - 1) . " client(s) connected to the server\n";
        $key = array_search($sock, $read);
        unset($read[$key]);
    }

    foreach ($read as $read_sock) {
        $data = @socket_read($read_sock, 1024, PHP_NORMAL_READ);
        $key = array_search($read_sock, $clients);

        if ($data === false) {
            unset($clients[$key]);
            echo "Client disconnected.\n";
            continue;
        }

        $data = trim($data);

        if (empty($data)) {
            continue;
        }

        $auth_re = '/^iam\:(.+)$/i';
        if (preg_match($auth_re, $data, $matches)) {
            $auth_clients[$key] = $matches[1];
            echo "Socket: $key authed; Username: {$matches[1]}\n";
            $data = "New user ({$matches[1]}) input into chat. Hello {$matches[1]}!!!";
        } else {
            $username = isset($auth_clients[$key]) ? $auth_clients[$key] : 'Anon' . $key;
            $data = "[$username]: $data";
        }
        
        foreach ($clients as $send_sock) {
            if ($send_sock == $sock)
                continue;
            else
                socket_write($send_sock, $data . "\n");
        }

    }
}
socket_close($sock);

File: python/socket_chat_server.py
# -*- coding: utf-8 -*-
"""
This is a simple example of socket module
"""
import socket
import select
import re

def run_server(ip, port):
    sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM, socket.SOL_TCP)
    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
    sock.bind((ip, port))
    sock.listen(1)

    clients = [sock]
    auth_clients = {}

    while 1:
        read = select.select(clients, [], [], 2)[0]

        if not read:
            continue

        if sock in read:
            new_sock = sock.accept()[0]
            clients.append(new_sock)
            new_sock.send("This simple chat server. Type 'iam:username' for authorize!\n")

            print "There are %s client(s) connected to the server" % (len(clients) -1)
            read.remove(sock)

        for read_sock in read:
            data = read_sock.recv(1024)
            if not data:
                print "Client disconected!"
                clients.remove(read_sock)
                continue

            data = data.strip()
            if not data:
                continue

            key = clients.index(read_sock)

            iam_usernames = re.findall(r"^iam:(.+)$", data)
            if iam_usernames:
                auth_clients[key] = iam_usernames[0]
                data = "New user (%s) input into chat. Hello (%s)!!!" % (iam_usernames[0], iam_usernames[0])
                print data
            else:
                if key in auth_clients:
                    username = auth_clients[key]
                else:
                    username = "Anon%s" % key

                data = "[%s]: %s" % (username, data)


            for write_sock in clients:
                if write_sock != sock:
                    write_sock.send(data + "\n")

    sock.close()

if __name__ == "__main__":
    run_server('', 9051)